adoc/chapters/cuda_backend.adoc

```diff
@@ -0,0 +1,598 @@
+three dimensions. SYCL sub-groups roughly map to
+cooperative groups `thread_block_tile` as it allows for the
+work-group/thread block to be further subdivided into concurrent threads.
+Likewise, thread blocks map directly to work-groups, and a
+single thread is a SYCL work-item.
+
+CUDA primarily synchronizes the threads through two functions,
+`cudaStreamSynchronize()` and `__syncthreads()`. 
+`cudaStreamSynchronize()` blocks work from being performed until all 
+threads on the device has been completed. `__syncthreads()` waits for 
+all threads within a thread block to reach the same point. So 
+`cudaStreamSynchronize()` is similar to queue.wait(), buffer 
+destruction, and other host-device synchronization events within SYCL. 
+`__syncthreads()` synchronizes the threads within a thread block which
+is analogous to the work-group barrier.
+
+CUDA's warp concept has no SYCL equivalent. If a user were to write 
+warp aware code it would be non-generic SYCL code and specific to the 
+CUDA backend.
+
+CUDA allows for more detailed thread and memory management through 
+Cooperative Groups. Cooperative Groups allow for synchronizing at the 
+grid level and organizing subgroups in sizes smaller than a warp. 
+Cooperative Groups do not have an equivalent within SYCL 2020 and are 
+not yet supported.
+
+==== Work Item Mapping
+
+A SYCL `nd_range` will tranpose indices as it maps to hardware memory.
+This gives better memory access patterns, in general.
+
+SYCL uses row major memory ordering, meaning in some memory object the 
+rows will be contiguous in memory. SYCL follows C++ convention in 
+this regard. Following the row-major paradigm, it is intuitive to imagine 
+each work-item in a `parallel_for` indexing through a contiguous block of 
+memory as it does its work. However, this gives poorly coalesced memory 
+accesses, as a given contiguous chunk of data being loaded may only pass 
+memory to a single work-item. More efficient memory access patterns are 
+achieved when each load of contiguous data can give data to as many 
+work-items as possible. Meaning the data used by a given work-item
+is non-contiguous.
+
+SYCL makes this intuitive row-major C++ approach give good memory access 
+patterns by flipping the indices of the `nd_range`, as it maps to hardware.
+
+The linear id (whose use is not recommended) of a two dimensional `nd_range` 
+can be calculated using:
+
+[source,c++]
+----
+cgh.parallel_for(range<2>(64, 128), [=](item<2> it) {
+  size_t linearIdx = it.get_id(1) + (it.get_id(0) * it.get_range(0));
+  ...
+});
+----
+
+Notice that rows appear to be accessed in a column-major, rather than 
+row-major, format. This is only the case because the indices are flipped 
+by the SYCL implementation. All memory in SYCL is stored in row-major format.
+
+It is best to avoid calculating the linear index manually; it is better
+to use a multi-dimensional `sycl::id` to index into memory, as it doesn't
+expose index-flipping to the user.
```

Comment:
@hdelan I also find it a bit unclear in parts. Perhaps you can clarify a little that the CUDA backend uses a different format to the SYCL frontend (if this is necessary/useful to mention here?) and best practice for a user to avoid running into related issues/ what they need to know about the index flipping if they can't avoid running into this issue?

Reply:
This has been dealt with in our internal draft, commit here will be made once other points are dealt with.